Wei/View/Send/SelectAmount/SelectAmountViewModel.swift

```diff
@@ -35,72 +37,124 @@ final class SelectAmountViewModel: InjectableViewModel {
         
         // User's usable fiat balance (totalBalance - txFee)
-        let availableBalance = Driver.combineLatest(fiatBalance, fiatTxFee) { balance, txFee -> Int64 in
-            let availableBalance = balance - txFee
+        let availableBalance = Driver.combineLatest(fiatBalance, fiatTxFee) { balance, txFee -> Fiat in
+            let availableBalance = balance.value - txFee.value
             // do not return negative value here.
-            return availableBalance > 0 ? availableBalance : 0
+            assert(availableBalance >= 0)
+            switch balance {
+            case .jpy:
+                return Fiat.jpy(availableBalance.toInt64())
+            case .usd:
+                return Fiat.usd(availableBalance)
+            }
         }
         
+        // inputAmount will be used to manage input text field text. it will prevents the text field from having
+        // invalid string value(which cannot be converted to Decimal).
+        let inputAmount = Driver
+            .combineLatest(input.amountTextFieldDidInput, availableBalance)
+            .map { inputAmount, availableBalance -> String in
+                
+                // If input text has more than one ".", then return the string with last letter dropped.
+                // eg, 12.9. -> 12.9
+                guard inputAmount.filter({ $0 == "." }).count <= 1 else {
+                    return String(inputAmount.dropLast())
+                }
+                
+                // If input text has more than 2 decimal points, return the string with last letter dropped
+                // eg, 1.324 -> 1.32
+                guard let subString = inputAmount.split(separator: ".").last, String(subString).count <= 2 else {
+                    return String(inputAmount.dropLast())
+                }
+                
+                // If input cannot be converted to Decimal type, then return the string with last letter dropped
+                // eg, 1.32a -> 1.32
+                guard let amount = Decimal(string: inputAmount) else {
+                    return String(inputAmount.dropLast())
+                }
+                
+                return amount <= availableBalance.value ?
+                    inputAmount : availableBalance.value.description
+            }
+        
         // fiatAmount represents an amount user gives as an input in text field.
         // if user's input is larger than user's fiat balance, it returns user's total balance.
         let inputFiatAmount = Driver
-            .combineLatest(input.amountTextFieldDidInput, availableBalance)
-            .map { inputAmount, balance -> Int64 in
-                guard let amount = Int64(inputAmount) else {
-                    return 0
+            .combineLatest(inputAmount, currency)
+            .map { amountText, currency -> Fiat in
+                let amount = Decimal(string: amountText) ?? 0
```

Comment:
does `0` have `toInt64()` method?

Reply:
Might it be `Decimal(0).toInt64()` ?

Reply:
```swift
Decimal(string: amountText) ?? 0
```

since the left side type is specified as `Decimal`, right side type will automatically be `Decimal` and `Decimal` has a method `toInt64()` so it's no problem 👍

Reply:
🥇